`make lint` fails with clang-tidy 8 (Ubuntu 19.04)

https://github.com/rootless-containers/slirp4netns/blob/v0.4.0-beta.1/main.c
```console
$ make lint
clang-tidy -warnings-as-errors='*' main.c slirp4netns.c api.c -- -I/usr/include/glib-2.0 -I/usr/lib/x86_64-linux-gnu/glib-2.0/include
2 warnings generated.
2 warnings generated.
2 warnings generated.
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:395:19: error: Potential memory leak [clang-analyzer-unix.Malloc,-warnings-as-errors]
    while ((opt = getopt_long(argc, argv, "ce:r:m:a:6hv", longopts, NULL)) !=
                  ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:622:5: note: Calling 'parse_args'
    parse_args(argc, argv, &options);
    ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:395:12: note: Assuming the condition is true
    while ((opt = getopt_long(argc, argv, "ce:r:m:a:6hv", longopts, NULL)) !=
           ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:395:5: note: Loop condition is true.  Entering loop body
    while ((opt = getopt_long(argc, argv, "ce:r:m:a:6hv", longopts, NULL)) !=
    ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:397:9: note: Control jumps to 'case 97:'  at line 453
        switch (opt) {
        ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:454:35: note: Memory is allocated
            options->api_socket = strdup(optarg);
                                  ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:455:13: note:  Execution continues on line 395
            break;
            ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:395:12: note: Assuming the condition is true
    while ((opt = getopt_long(argc, argv, "ce:r:m:a:6hv", longopts, NULL)) !=
           ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:395:5: note: Loop condition is true.  Entering loop body
    while ((opt = getopt_long(argc, argv, "ce:r:m:a:6hv", longopts, NULL)) !=
    ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:397:9: note: Control jumps to 'case 97:'  at line 453
        switch (opt) {
        ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:455:13: note:  Execution continues on line 395
            break;
            ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:395:19: note: Potential memory leak
    while ((opt = getopt_long(argc, argv, "ce:r:m:a:6hv", longopts, NULL)) !=
                  ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:446:17: error: Potential memory leak [clang-analyzer-unix.Malloc,-warnings-as-errors]
            if (access(options->userns_path, F_OK) == -1) {
                ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:622:5: note: Calling 'parse_args'
    parse_args(argc, argv, &options);
    ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:395:12: note: Assuming the condition is true
    while ((opt = getopt_long(argc, argv, "ce:r:m:a:6hv", longopts, NULL)) !=
           ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:395:5: note: Loop condition is true.  Entering loop body
    while ((opt = getopt_long(argc, argv, "ce:r:m:a:6hv", longopts, NULL)) !=
    ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:397:9: note: Control jumps to 'case -45:'  at line 444
        switch (opt) {
        ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:445:36: note: Memory is allocated
            options->userns_path = strdup(optarg);
                                   ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:446:17: note: Assuming the condition is false
            if (access(options->userns_path, F_OK) == -1) {
                ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:446:13: note: Taking false branch
            if (access(options->userns_path, F_OK) == -1) {
            ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:452:13: note:  Execution continues on line 395
            break;
            ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:395:12: note: Assuming the condition is true
    while ((opt = getopt_long(argc, argv, "ce:r:m:a:6hv", longopts, NULL)) !=
           ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:395:5: note: Loop condition is true.  Entering loop body
    while ((opt = getopt_long(argc, argv, "ce:r:m:a:6hv", longopts, NULL)) !=
    ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:397:9: note: Control jumps to 'case -45:'  at line 444
        switch (opt) {
        ^
/home/suda/gopath/src/github.com/rootless-containers/slirp4netns/main.c:446:17: note: Potential memory leak
            if (access(options->userns_path, F_OK) == -1) {
                ^
2 warnings treated as errors
make: *** [Makefile:2133: lint] Error 2
```

```console
$ clang-tidy --version
LLVM (http://llvm.org/):
  LLVM version 8.0.0

  Optimized build.
  Default target: x86_64-pc-linux-gnu
  Host CPU: skylake
```

--- a/main.c
+++ b/main.c
@@ -66,8 +66,6 @@ static int nsenter(pid_t target_pid, char *netns, char *userns)
         return -1;
     }
     close(netnsfd);
-    free(netns);
-    free(userns);
     return 0;
 }
 
@@ -358,9 +356,14 @@ static void options_destroy(struct options *options)
         free(options->netns_type);
         options->netns_type = NULL;
     }
-    // options->netns_path and options->userns_path are
-    // getting freed in the nsenter function
-    // options itself is not freed, because it can be on the stack.
+    if (options->netns_path != NULL) {
+        free(options->netns_path);
+        options->netns_path = NULL;
+    }
+    if (options->userns_path != NULL) {
+        free(options->userns_path);
+        options->userns_path = NULL;
+    }
 }
 
 // * caller does not need to call options_init()
@@ -403,8 +406,7 @@ static void parse_args(int argc, char *const argv[], struct options *options)
             options->exit_fd = strtol(optarg, NULL, 10);
             if (errno) {
                 perror("strtol");
-                usage(argv[0]);
-                exit(EXIT_FAILURE);
+                goto error;
             }
             break;
         case 'r':
@@ -412,8 +414,7 @@ static void parse_args(int argc, char *const argv[], struct options *options)
             options->ready_fd = strtol(optarg, NULL, 10);
             if (errno) {
                 perror("strtol");
-                usage(argv[0]);
-                exit(EXIT_FAILURE);
+                goto error;
             }
             break;
         case 'm':
@@ -421,8 +422,7 @@ static void parse_args(int argc, char *const argv[], struct options *options)
             options->mtu = strtol(optarg, NULL, 10);
             if (errno) {
                 perror("strtol");
-                usage(argv[0]);
-                exit(EXIT_FAILURE);
+                goto error;
             }
             break;
         case CIDR:
@@ -446,8 +446,7 @@ static void parse_args(int argc, char *const argv[], struct options *options)
             if (access(options->userns_path, F_OK) == -1) {
                 fprintf(stderr, "userns path doesn't exist: %s\n",
                         options->userns_path);
-                usage(argv[0]);
-                exit(EXIT_FAILURE);
+                goto error;
             }
             break;
         case 'a':
@@ -460,12 +459,14 @@ static void parse_args(int argc, char *const argv[], struct options *options)
         case 'h':
             usage(argv[0]);
             exit(EXIT_SUCCESS);
+            break;
         case 'v':
             version();
             exit(EXIT_SUCCESS);
+            break;
         default:
-            usage(argv[0]);
-            exit(EXIT_FAILURE);
+            goto error;
+            break;
         }
     }
 #undef CIDR
@@ -475,11 +476,10 @@ static void parse_args(int argc, char *const argv[], struct options *options)
 #undef _DEPRECATED_NO_HOST_LOOPBACK
     if (options->mtu == 0 || options->mtu > 65521) {
         fprintf(stderr, "MTU must be a positive integer (< 65522)\n");
-        exit(EXIT_FAILURE);
+        goto error;
     }
     if (argc - optind < 2) {
-        usage(argv[0]);
-        exit(EXIT_FAILURE);
+        goto error;
     }
     if (!options->netns_type ||
         strcmp(options->netns_type, DEFAULT_NETNS_TYPE) == 0) {
@@ -487,18 +487,21 @@ static void parse_args(int argc, char *const argv[], struct options *options)
         options->target_pid = strtol(argv[optind], NULL, 10);
         if (errno != 0) {
             perror("strtol");
-            usage(argv[0]);
-            exit(EXIT_FAILURE);
+            goto error;
         }
     } else {
         options->netns_path = strdup(argv[optind]);
         if (access(options->netns_path, F_OK) == -1) {
             perror("existing path expected when --netns-type=path");
-            usage(argv[0]);
-            exit(EXIT_FAILURE);
+            goto error;
         }
     }
     options->tapname = strdup(argv[optind + 1]);
+    return;
+error:
+    usage(argv[0]);
+    options_destroy(options);
+    exit(EXIT_FAILURE);
 }
 
 static int from_regmatch(char *buf, size_t buf_len, regmatch_t match,

--- a/main.c
+++ b/main.c
@@ -372,6 +372,10 @@ static void options_destroy(struct options *options)
 static void parse_args(int argc, char *const argv[], struct options *options)
 {
     int opt;
+    char *optarg_cidr = NULL;
+    char *optarg_netns_type = NULL;
+    char *optarg_userns_path = NULL;
+    char *optarg_api_socket = NULL;
 #define CIDR -42
 #define DISABLE_HOST_LOOPBACK -43
 #define NETNS_TYPE -44
@@ -395,6 +399,7 @@ static void parse_args(int argc, char *const argv[], struct options *options)
         { 0, 0, 0, 0 },
     };
     options_init(options);
+    /* NOTE: clang-tidy hates strdup(optarg) in the while loop (#112) */
     while ((opt = getopt_long(argc, argv, "ce:r:m:a:6hv", longopts, NULL)) !=
            -1) {
         switch (opt) {
@@ -426,7 +431,7 @@ static void parse_args(int argc, char *const argv[], struct options *options)
             }
             break;
         case CIDR:
-            options->cidr = strdup(optarg);
+            optarg_cidr = optarg;
             break;
         case _DEPRECATED_NO_HOST_LOOPBACK:
             // There was no tagged release with support for --no-host-loopback.
@@ -439,18 +444,18 @@ static void parse_args(int argc, char *const argv[], struct options *options)
             options->disable_host_loopback = true;
             break;
         case NETNS_TYPE:
-            options->netns_type = strdup(optarg);
+            optarg_netns_type = optarg;
             break;
         case USERNS_PATH:
-            options->userns_path = strdup(optarg);
-            if (access(options->userns_path, F_OK) == -1) {
+            optarg_userns_path = optarg;
+            if (access(optarg_userns_path, F_OK) == -1) {
                 fprintf(stderr, "userns path doesn't exist: %s\n",
-                        options->userns_path);
+                        optarg_userns_path);
                 goto error;
             }
             break;
         case 'a':
-            options->api_socket = strdup(optarg);
+            optarg_api_socket = optarg;
             break;
         case '6':
             options->enable_ipv6 = true;
@@ -468,6 +473,18 @@ static void parse_args(int argc, char *const argv[], struct options *options)
             goto error;
             break;
         }
+    }
+    if (optarg_cidr != NULL) {
+        options->cidr = strdup(optarg_cidr);
+    }
+    if (optarg_netns_type != NULL) {
+        options->netns_type = strdup(optarg_netns_type);
+    }
+    if (optarg_userns_path != NULL) {
+        options->userns_path = strdup(optarg_userns_path);
+    }
+    if (optarg_api_socket != NULL) {
+        options->api_socket = strdup(optarg_api_socket);
     }
 #undef CIDR
 #undef DISABLE_HOST_LOOPBACK

--- a/main.c
+++ b/main.c
@@ -372,6 +372,7 @@ static void options_destroy(struct options *options)
 static void parse_args(int argc, char *const argv[], struct options *options)
 {
     int opt;
+    char *strtol_e = NULL;
     char *optarg_cidr = NULL;
     char *optarg_netns_type = NULL;
     char *optarg_userns_path = NULL;
@@ -408,25 +409,26 @@ static void parse_args(int argc, char *const argv[], struct options *options)
             break;
         case 'e':
             errno = 0;
-            options->exit_fd = strtol(optarg, NULL, 10);
-            if (errno) {
-                perror("strtol");
+            options->exit_fd = strtol(optarg, &strtol_e, 10);
+            if (errno || *strtol_e != '\0' || options->exit_fd < 0) {
+                fprintf(stderr, "exit-fd must be a non-negative integer\n");
                 goto error;
             }
             break;
         case 'r':
             errno = 0;
-            options->ready_fd = strtol(optarg, NULL, 10);
-            if (errno) {
-                perror("strtol");
+            options->ready_fd = strtol(optarg, &strtol_e, 10);
+            if (errno || *strtol_e != '\0' || options->ready_fd < 0) {
+                fprintf(stderr, "ready-fd must be a non-negative integer\n");
                 goto error;
             }
             break;
         case 'm':
             errno = 0;
-            options->mtu = strtol(optarg, NULL, 10);
-            if (errno) {
-                perror("strtol");
+            options->mtu = strtol(optarg, &strtol_e, 10);
+            if (errno || *strtol_e != '\0' || options->mtu <= 0 ||
+                options->mtu > 65521) {
+                fprintf(stderr, "MTU must be a positive integer (< 65522)\n");
                 goto error;
             }
             break;
@@ -491,19 +493,15 @@ static void parse_args(int argc, char *const argv[], struct options *options)
 #undef NETNS_TYPE
 #undef USERNS_PATH
 #undef _DEPRECATED_NO_HOST_LOOPBACK
-    if (options->mtu == 0 || options->mtu > 65521) {
-        fprintf(stderr, "MTU must be a positive integer (< 65522)\n");
-        goto error;
-    }
     if (argc - optind < 2) {
         goto error;
     }
     if (!options->netns_type ||
         strcmp(options->netns_type, DEFAULT_NETNS_TYPE) == 0) {
         errno = 0;
-        options->target_pid = strtol(argv[optind], NULL, 10);
-        if (errno != 0) {
-            perror("strtol");
+        options->target_pid = strtol(argv[optind], &strtol_e, 10);
+        if (errno || *strtol_e != '\0' || options->target_pid <= 0) {
+            fprintf(stderr, "PID must be a positive integer\n");
             goto error;
         }
     } else {

